In [1]:
import os
import sys

PROJECT_ROOT = os.path.abspath("..")

if PROJECT_ROOT not in sys.path:
    sys.path.append(PROJECT_ROOT)

from src.data import load_prices

In [2]:
prices = load_prices("../data/raw/prices.csv")

prices.head()

,AAPL,AMZN,BAC,CVX,GOOGL,HD,JNJ,JPM,KO,MA,MCD,META,MSFT,NVDA,PEP,PG,SBUX,V,WMT,XOM
Date,,,,,,,,,,,,,,,,,,,,
2015-01-02,24.192604,15.4260,14.050793,68.898338,26.244781,78.780350,76.110909,46.066780,29.390274,79.714676,69.804466,77.767075,39.681744,0.482423,66.582924,65.555763,32.260994,61.088104,23.127542,57.145554
2015-01-05,23.511065,15.1095,13.642614,66.144341,25.744715,77.127518,75.579308,44.636642,29.390274,77.472473,69.033524,76.518066,39.316833,0.474275,66.082375,65.244041,31.643021,59.739658,23.060234,55.581917
2015-01-06,23.513273,14.7645,13.234435,66.113792,25.109354,76.891373,75.207924,43.479275,29.613434,77.305023,69.160774,75.487099,38.739761,0.459895,65.581764,64.946877,31.385550,59.354717,23.237928,55.286449
2015-01-07,23.842978,14.9210,13.297234,66.058678,25.035503,79.526802,76.868225,43.545609,29.983086,78.507484,70.365829,75.487099,39.231968,0.458697,67.499466,65.287560,32.158001,60.149960,23.854486,55.846645
2015-01-08,24.759089,15.0230,13.571970,67.570320,25.122730,81.286270,77.472626,44.518707,30.345755,79.728645,70.627785,77.499435,40.386089,0.475952,68.726227,66.034149,32.676929,60.956718,24.357960,56.776211


In [3]:
from src.returns import (
    calculate_returns,
    annualized_returns,
    annualized_covariance,
    annualized_volatility
)

In [4]:
returns = calculate_returns(
    prices
)

returns.head()

,AAPL,AMZN,BAC,CVX,GOOGL,HD,JNJ,JPM,KO,MA,MCD,META,MSFT,NVDA,PEP,PG,SBUX,V,WMT,XOM
Date,,,,,,,,,,,,,,,,,,,,
2015-01-05,-0.028171,-0.020517,-0.029050,-0.039972,-0.019054,-0.020980,-0.006985,-0.031045,0.000000,-0.028128,-0.011044,-0.016061,-0.009196,-0.016890,-0.007518,-0.004755,-0.019155,-0.022074,-0.002910,-0.027362
2015-01-06,0.000094,-0.022833,-0.029919,-0.000462,-0.024679,-0.003062,-0.004914,-0.025929,0.007593,-0.002161,0.001843,-0.013474,-0.014677,-0.030319,-0.007576,-0.004555,-0.008137,-0.006444,0.007706,-0.005316
2015-01-07,0.014022,0.010600,0.004745,-0.000834,-0.002941,0.034275,0.022076,0.001526,0.012483,0.015555,0.017424,0.000000,0.012705,-0.002605,0.029241,0.005246,0.024612,0.013398,0.026532,0.010133
2015-01-08,0.038423,0.006836,0.020661,0.022883,0.003484,0.022124,0.007863,0.022347,0.012096,0.015555,0.003723,0.026658,0.029418,0.037617,0.018174,0.011435,0.016137,0.013412,0.021106,0.016645
2015-01-09,0.001072,-0.011749,-0.017930,-0.019926,-0.012211,-0.017148,-0.013629,-0.017387,-0.011032,-0.012744,-0.012187,-0.005628,-0.008405,0.004029,-0.006770,-0.009331,-0.032731,-0.014823,-0.012380,-0.001410


In [5]:
expected_returns = annualized_returns(
    returns
)

expected_returns.sort_values(
    ascending=False
)

NVDA     0.661567
AMZN     0.301402
GOOGL    0.267700
META     0.266818
MSFT     0.264488
AAPL     0.262454
MA       0.214824
JPM      0.213234
V        0.189156
BAC      0.171143
WMT      0.166545
HD       0.163455
MCD      0.154723
SBUX     0.128390
CVX      0.113381
JNJ      0.106783
XOM      0.104781
KO       0.093863
PG       0.087585
PEP      0.086816
dtype: float64

In [6]:
volatility = annualized_volatility(
    returns
)

volatility.sort_values()

KO       0.177620
JNJ      0.182034
PG       0.185158
PEP      0.190351
MCD      0.202526
WMT      0.215145
HD       0.242577
V        0.242624
MA       0.263642
MSFT     0.268916
JPM      0.271535
XOM      0.274216
SBUX     0.286887
GOOGL    0.288466
AAPL     0.288506
CVX      0.291472
BAC      0.307520
AMZN     0.329333
META     0.375375
NVDA     0.487002
dtype: float64

In [7]:
cov_matrix = annualized_covariance(
    returns
)

cov_matrix.head()

,AAPL,AMZN,BAC,CVX,GOOGL,HD,JNJ,JPM,KO,MA,MCD,META,MSFT,NVDA,PEP,PG,SBUX,V,WMT,XOM
AAPL,0.083235,0.052722,0.037210,0.028461,0.049089,0.034615,0.016624,0.033207,0.017502,0.043793,0.022419,0.056518,0.051453,0.074122,0.021829,0.018913,0.037574,0.039123,0.019963,0.024407
AMZN,0.052722,0.108460,0.032443,0.020126,0.060122,0.032398,0.010833,0.028072,0.010620,0.040966,0.016570,0.073937,0.057678,0.082429,0.014071,0.012264,0.036121,0.036255,0.017928,0.016797
BAC,0.037210,0.032443,0.094568,0.050971,0.035247,0.035947,0.018941,0.073867,0.020745,0.044977,0.025166,0.037858,0.035444,0.049824,0.018998,0.017351,0.040550,0.041682,0.016586,0.045838
CVX,0.028461,0.020126,0.050971,0.084956,0.025949,0.028139,0.018023,0.046159,0.019947,0.035949,0.023471,0.025577,0.026954,0.035758,0.017804,0.013987,0.030846,0.032669,0.012309,0.066710
GOOGL,0.049089,0.060122,0.035247,0.025949,0.083213,0.030344,0.014187,0.031730,0.014920,0.041084,0.019787,0.065952,0.053082,0.073369,0.016711,0.014426,0.034651,0.037391,0.015555,0.021431


In [8]:
######### Equal Weight Portfolio
from src.optimization import (
    equal_weight_portfolio,
    portfolio_return,
    portfolio_volatility,
    portfolio_sharpe
)

In [9]:
weights = equal_weight_portfolio(
    returns
)

weights

array([0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05,
       0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05])

In [10]:
portfolio_return(
    weights,
    expected_returns
)

np.float64(0.20095533476403982)

In [11]:
portfolio_volatility(
    weights,
    cov_matrix
)

np.float64(0.17813237275787472)

In [12]:
portfolio_sharpe(
    weights,
    expected_returns,
    cov_matrix
)

np.float64(1.015847551809138)

| Metric                 |     Result |
| ---------------------- | ---------: |
| Expected Annual Return | **20.10%** |
| Annual Volatility      | **17.81%** |
| Sharpe Ratio (Rf = 2%) |   **1.02** |


Interpretation
Return ≈ 20%

This is high, yet reasonable for the 2015–2025 period, especially given that the investment universe includes:

NVDA
META
MSFT
GOOGL
AAPL

...which experienced an exceptional decade.

Volatility ≈ 17.8%

This is typical for a diversified equity portfolio.

Sharpe ≈ 1.02

A Sharpe ratio > 1 is considered very good.

Thus, the benchmark portfolio already demonstrates a strong risk-return profile.

###### Minimum Variance Portfolio

In [13]:
from src.optimization import minimum_variance_portfolio

In [14]:
min_var_weights = minimum_variance_portfolio(
    cov_matrix
)

min_var_weights

array([5.01953678e-18, 5.34227784e-02, 1.32618775e-17, 5.71582195e-18,
       1.65989942e-02, 3.64228530e-18, 2.39706248e-01, 9.61894977e-18,
       1.98267225e-01, 0.00000000e+00, 1.30155503e-01, 3.23559050e-03,
       7.73244644e-18, 5.22501996e-19, 6.21048920e-18, 1.28201188e-01,
       3.30681663e-18, 6.85080248e-18, 1.65511681e-01, 6.49007927e-02])

In [15]:
portfolio_return(
    min_var_weights,
    expected_returns
)

np.float64(0.13134695946132507)

In [16]:
portfolio_volatility(
    min_var_weights,
    cov_matrix
)

np.float64(0.13960970991352184)

In [17]:
portfolio_sharpe(
    min_var_weights,
    expected_returns,
    cov_matrix
)

np.float64(0.7975588483802201)

In [18]:
import pandas as pd
pd.Series(
    min_var_weights,
    index=returns.columns
).sort_values(
    ascending=False
)

JNJ      2.397062e-01
KO       1.982672e-01
WMT      1.655117e-01
MCD      1.301555e-01
PG       1.282012e-01
XOM      6.490079e-02
AMZN     5.342278e-02
GOOGL    1.659899e-02
META     3.235590e-03
BAC      1.326188e-17
JPM      9.618950e-18
MSFT     7.732446e-18
V        6.850802e-18
PEP      6.210489e-18
CVX      5.715822e-18
AAPL     5.019537e-18
HD       3.642285e-18
SBUX     3.306817e-18
NVDA     5.225020e-19
MA       0.000000e+00
dtype: float64

In [19]:
from src.performance import (
    portfolio_metrics,
    create_comparison_table
)

In [20]:
equal_weight_result = portfolio_metrics(
    "Equal Weight",
    weights,
    expected_returns,
    cov_matrix
)

In [21]:
min_var_weights

array([5.01953678e-18, 5.34227784e-02, 1.32618775e-17, 5.71582195e-18,
       1.65989942e-02, 3.64228530e-18, 2.39706248e-01, 9.61894977e-18,
       1.98267225e-01, 0.00000000e+00, 1.30155503e-01, 3.23559050e-03,
       7.73244644e-18, 5.22501996e-19, 6.21048920e-18, 1.28201188e-01,
       3.30681663e-18, 6.85080248e-18, 1.65511681e-01, 6.49007927e-02])

In [22]:
min_var_result = portfolio_metrics(
    "Minimum Variance",
    min_var_weights,
    expected_returns,
    cov_matrix
)

In [23]:
results = [
    equal_weight_result,
    min_var_result
]


comparison = create_comparison_table(
    results
)


comparison

,Portfolio,Annual Return,Volatility,Sharpe Ratio
0,Equal Weight,0.200955,0.178132,1.015848
1,Minimum Variance,0.131347,0.139610,0.797559


##### Maximum Sharpe Portfolio

In [24]:
from src.optimization import (
    maximum_sharpe_portfolio
)

In [25]:
max_sharpe_weights = maximum_sharpe_portfolio(
    expected_returns,
    cov_matrix
)

In [26]:
pd.Series(
    max_sharpe_weights,
    index=returns.columns
).sort_values(
    ascending=False
)

NVDA     3.844070e-01
WMT      2.474031e-01
MCD      1.642074e-01
JNJ      8.754962e-02
JPM      5.803583e-02
AMZN     5.631999e-02
GOOGL    2.077044e-03
SBUX     1.023937e-16
BAC      1.011373e-16
MA       7.469579e-17
AAPL     7.038695e-17
HD       4.832658e-17
PG       2.692152e-17
META     1.098742e-17
MSFT     4.037970e-18
CVX      0.000000e+00
KO       0.000000e+00
PEP      0.000000e+00
V        0.000000e+00
XOM      0.000000e+00
dtype: float64

In [27]:
max_sharpe_result = portfolio_metrics(
    "Maximum Sharpe",
    max_sharpe_weights,
    expected_returns,
    cov_matrix
)


results.append(
    max_sharpe_result
)


comparison = create_comparison_table(
    results
)


comparison

,Portfolio,Annual Return,Volatility,Sharpe Ratio
0,Equal Weight,0.200955,0.178132,1.015848
1,Minimum Variance,0.131347,0.139610,0.797559
2,Maximum Sharpe,0.360176,0.240069,1.416996


#### Risk Parity Portfolio

In [28]:
from src.optimization import (
    risk_parity_portfolio
)

In [29]:
risk_parity_weights = risk_parity_portfolio(
    cov_matrix
)

In [30]:
pd.Series(
    risk_parity_weights,
    index=returns.columns
).sort_values(
    ascending=False
)

JNJ      0.072368
WMT      0.071704
PG       0.071480
KO       0.070446
PEP      0.068445
MCD      0.064941
XOM      0.054348
HD       0.050598
CVX      0.046144
SBUX     0.044025
V        0.043809
JPM      0.042179
GOOGL    0.041557
AMZN     0.040548
MSFT     0.040043
AAPL     0.039189
MA       0.038140
BAC      0.036061
META     0.034218
NVDA     0.029757
dtype: float64

In [31]:
risk_parity_result = portfolio_metrics(
    "Risk Parity",
    risk_parity_weights,
    expected_returns,
    cov_matrix
)


results.append(
    risk_parity_result
)


comparison = create_comparison_table(
    results
)


comparison

,Portfolio,Annual Return,Volatility,Sharpe Ratio
0,Equal Weight,0.200955,0.178132,1.015848
1,Minimum Variance,0.131347,0.139610,0.797559
2,Maximum Sharpe,0.360176,0.240069,1.416996
3,Risk Parity,0.178066,0.164438,0.961253


#### Efficient Frontier

In [32]:
from src.efficient_frontier import (
    generate_random_portfolios
)

In [33]:
frontier = generate_random_portfolios(
    expected_returns,
    cov_matrix,
    n_portfolios=5000
)

In [34]:
frontier.head()

,Return,Volatility,Sharpe,Weights
0,0.215131,0.177973,1.096405,"[0.06428473041091327, 0.025643761748117146, 0...."
1,0.220613,0.186921,1.073251,"[0.10318898360022849, 0.032393244195149794, 0...."
2,0.199502,0.171996,1.043639,"[0.054210544002450345, 0.022096637594065414, 0..."
3,0.203866,0.180151,1.020621,"[0.02066807280932882, 0.027849204991585105, 0...."
4,0.182951,0.168124,0.969228,"[0.03973751853924116, 0.03994969241556111, 0.0..."


In [35]:
frontier.loc[
    frontier["Sharpe"].idxmax()
]

Return                                                 0.256991
Volatility                                              0.20007
Sharpe                                                 1.184541
Weights       [0.0433508867337532, 0.09533812896977617, 0.00...
Name: 3231, dtype: object

In [36]:
from src.plotting import plot_efficient_frontier

In [37]:
portfolio_points = [
    {
        "Portfolio":"Equal Weight",
        "Return":portfolio_return(
            weights,
            expected_returns
        ),
        "Volatility":portfolio_volatility(
            weights,
            cov_matrix
        )
    },

    {
        "Portfolio":"Minimum Variance",
        "Return":portfolio_return(
            min_var_weights,
            expected_returns
        ),
        "Volatility":portfolio_volatility(
            min_var_weights,
            cov_matrix
        )
    },

    {
        "Portfolio":"Maximum Sharpe",
        "Return":portfolio_return(
            max_sharpe_weights,
            expected_returns
        ),
        "Volatility":portfolio_volatility(
            max_sharpe_weights,
            cov_matrix
        )
    },

    {
        "Portfolio":"Risk Parity",
        "Return":portfolio_return(
            risk_parity_weights,
            expected_returns
        ),
        "Volatility":portfolio_volatility(
            risk_parity_weights,
            cov_matrix
        )
    }
]

In [38]:
plot_efficient_frontier(
    frontier,
    portfolio_points
)

##### Risk Analytics Module

In [39]:
from src.risk import (
    portfolio_returns,
    cumulative_returns,
    maximum_drawdown,
    rolling_volatility
)

In [40]:
max_sharpe_daily = portfolio_returns(
    returns,
    max_sharpe_weights
)

In [41]:
equity_curve = cumulative_returns(
    max_sharpe_daily
)

equity_curve.tail()

Date
2025-12-23    37.910609
2025-12-24    38.030582
2025-12-26    38.127836
2025-12-29    37.940162
2025-12-30    37.818154
dtype: float64

In [42]:
maximum_drawdown(
    equity_curve
)

np.float64(-0.3582355511080442)

In [43]:
rolling_vol = rolling_volatility(
    max_sharpe_daily
)

rolling_vol.tail()

Date
2025-12-23    0.135105
2025-12-24    0.133415
2025-12-26    0.133144
2025-12-29    0.133680
2025-12-30    0.133936
dtype: float64

In [44]:
maximum_drawdown(equity_curve)

np.float64(-0.3582355511080442)

In [45]:
from src.performance import portfolio_risk_metrics

In [46]:
risk_results = []

In [47]:
risk_results.append(
    portfolio_risk_metrics(
        "Equal Weight",
        returns,
        weights,
        expected_returns,
        cov_matrix
    )
)

risk_results.append(
    portfolio_risk_metrics(
        "Minimum Variance",
        returns,
        min_var_weights,
        expected_returns,
        cov_matrix
    )
)

risk_results.append(
    portfolio_risk_metrics(
        "Maximum Sharpe",
        returns,
        max_sharpe_weights,
        expected_returns,
        cov_matrix
    )
)


risk_results.append(
    portfolio_risk_metrics(
        "Risk Parity",
        returns,
        risk_parity_weights,
        expected_returns,
        cov_matrix
    )
)


risk_summary = pd.DataFrame(
    risk_results
)

risk_summary

,Portfolio,Annual Return,Volatility,Sharpe Ratio,Max Drawdown
0,Equal Weight,0.200955,0.178132,1.015848,-0.328736
1,Minimum Variance,0.131347,0.139610,0.797559,-0.264315
2,Maximum Sharpe,0.360176,0.240069,1.416996,-0.358236
3,Risk Parity,0.178066,0.164438,0.961253,-0.317561


In [48]:
from src.risk import (
    portfolio_returns,
    cumulative_returns
)

In [49]:
from src.plotting import (
    plot_cumulative_returns,
    plot_drawdown
)

In [50]:
portfolio_daily = {}

portfolio_daily["Equal Weight"] = portfolio_returns(
    returns,
    weights
)

portfolio_daily["Minimum Variance"] = portfolio_returns(
    returns,
    min_var_weights
)

portfolio_daily["Maximum Sharpe"] = portfolio_returns(
    returns,
    max_sharpe_weights
)

portfolio_daily["Risk Parity"] = portfolio_returns(
    returns,
    risk_parity_weights
)

In [51]:
portfolio_daily = pd.DataFrame(
    portfolio_daily
)

In [52]:
cumulative = (
    1 + portfolio_daily
).cumprod()


plot_cumulative_returns(
    cumulative
)

In [53]:
drawdown = (
    cumulative /
    cumulative.cummax()
) - 1


plot_drawdown(
    drawdown
)

In [54]:
risk_summary

,Portfolio,Annual Return,Volatility,Sharpe Ratio,Max Drawdown
0,Equal Weight,0.200955,0.178132,1.015848,-0.328736
1,Minimum Variance,0.131347,0.139610,0.797559,-0.264315
2,Maximum Sharpe,0.360176,0.240069,1.416996,-0.358236
3,Risk Parity,0.178066,0.164438,0.961253,-0.317561


In [55]:
frontier

,Return,Volatility,Sharpe,Weights
0,0.215131,0.177973,1.096405,"[0.06428473041091327, 0.025643761748117146, 0...."
1,0.220613,0.186921,1.073251,"[0.10318898360022849, 0.032393244195149794, 0...."
2,0.199502,0.171996,1.043639,"[0.054210544002450345, 0.022096637594065414, 0..."
3,0.203866,0.180151,1.020621,"[0.02066807280932882, 0.027849204991585105, 0...."
4,0.182951,0.168124,0.969228,"[0.03973751853924116, 0.03994969241556111, 0.0..."
...,...,...,...,...
4995,0.216989,0.185951,1.059359,"[0.008653224757090168, 0.08544398530967538, 0...."
4996,0.192315,0.179758,0.958593,"[0.044944328249800276, 0.018323477601495157, 0..."
4997,0.206958,0.183609,1.018238,"[0.04398517029439639, 0.036506314514737105, 0...."
4998,0.197349,0.177720,0.997911,"[0.023347178211668377, 0.06252831866040613, 0...."


In [56]:
portfolio_points

[{'Portfolio': 'Equal Weight',
  'Return': np.float64(0.20095533476403982),
  'Volatility': np.float64(0.17813237275787472)},
 {'Portfolio': 'Minimum Variance',
  'Return': np.float64(0.13134695946132507),
  'Volatility': np.float64(0.13960970991352184)},
 {'Portfolio': 'Maximum Sharpe',
  'Return': np.float64(0.360176450794066),
  'Volatility': np.float64(0.2400687543603501)},
 {'Portfolio': 'Risk Parity',
  'Return': np.float64(0.17806598931718773),
  'Volatility': np.float64(0.16443750967051154)}]

In [57]:
cumulative

,Equal Weight,Minimum Variance,Maximum Sharpe,Risk Parity
Date,,,,
2015-01-05,0.981957,0.992557,0.987365,0.983933
2015-01-06,0.972764,0.991803,0.974808,0.976797
2015-01-07,0.984836,1.007943,0.985566,0.990396
2015-01-08,1.002888,1.019350,1.007910,1.007266
2015-01-09,0.991239,1.007907,1.001454,0.995436
...,...,...,...,...
2025-12-23,7.590246,3.778641,37.910609,6.056162
2025-12-24,7.622398,3.802187,38.030582,6.084074
2025-12-26,7.620321,3.796040,38.127836,6.081247


In [58]:
drawdown

,Equal Weight,Minimum Variance,Maximum Sharpe,Risk Parity
Date,,,,
2015-01-05,0.000000,0.000000,0.000000,0.000000
2015-01-06,-0.009361,-0.000759,-0.012718,-0.007252
2015-01-07,0.000000,0.000000,-0.001822,0.000000
2015-01-08,0.000000,0.000000,0.000000,0.000000
2015-01-09,-0.011616,-0.011226,-0.006405,-0.011744
...,...,...,...,...
2025-12-23,0.000000,-0.022191,0.000000,-0.006085
2025-12-24,0.000000,-0.016098,0.000000,-0.001505
2025-12-26,-0.000272,-0.017689,0.000000,-0.001969
